### PR TITLE
Add ParentOrElse Sampler

### PR DESCRIPTION
**Pull request by MitchellDumovic**
Introduce a new kind of composite sampler - a `ParentOrElseSampler` - which is specified in the spec document [here](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#built-in-samplers). A `ParentOrElseSampler` is initialized with a sampler delegate and makes the same sampling decision as the parent. If there is no parent, it uses the provided delegate to make a decision. 

There is a separate question as to whether or not this sampler (with an AlwaysOn delegate) should be used as the default Sampler, which would fix the "bug" outlined in https://github.com/open-telemetry/opentelemetry-java/issues/1445

### CODE REVIEW

**Comment by bogdandrutu** [comment]
@MitchellDumovic see https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#built-in-samplers the AlwaysOn does what specs suggest

File: sdk/src/main/java/io/opentelemetry/sdk/trace/Samplers.java [hunk 1]
```diff
@@ -176,6 +189,32 @@ public String getDescription() {
     }
   }
 
+  @Immutable
+  private enum FollowParentSampler implements Sampler {
+    INSTANCE;
+
+    // If a parent is set, always follows the same sampling decision as the parent.
+    // Otherwise, always makes a "yes" decision on {@link Span} sampling.
```
Comment (anuraaga) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/Samplers.java:197]: It seems like follow parent should be a setting on any sampler, or we should support multiple samplers where any sampler can return something like `NO_DECISION` to allow another sampler to pick it up. Always yes seems to be unrelated to following the parent, right?

Comment (anuraaga) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/Samplers.java:197]: Oops just jumped through bogdan's link, yeah shouldn't we use parentorelse instead of adding this sampler?

I would argue ParentOrElse(AlwaysOn) should be the default for the SDK though

File: sdk/src/main/java/io/opentelemetry/sdk/trace/config/TraceConfig.java [hunk 2]
```diff
@@ -82,8 +82,7 @@
 @Immutable
 public abstract class TraceConfig {
   // These values are the default values for all the global parameters.
-  // TODO: decide which default sampler to use
-  private static final Sampler DEFAULT_SAMPLER = Samplers.alwaysOn();
+  private static final Sampler DEFAULT_SAMPLER = Samplers.parentOrElse(Samplers.alwaysOn());
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/config/TraceConfig.java:85]: I know I am annoying, but can we do this in a followup PR? This is an important change and would like to make sure everyone aggress on, just adding the parentElse is trivial to be merged.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/Samplers.java (hunk 1 above)
Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/Samplers.java]: @anuraaga can we argue in a separate PR :)

File: sdk/src/main/java/io/opentelemetry/sdk/trace/config/TraceConfig.java (hunk 2 above)
Comment (MitchellDumovic) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/config/TraceConfig.java:85]: That's totally fine! I hadn't realized initially that the default was already set in the spec - I can also create an issue in the spec repo to discuss this instead of doing it in a followup PR here if that makes more sense?

**Comment by tedsuo** [comment]
A change like this should be standardized and recorded in the spec, correct?

**Comment by bogdandrutu** [comment]
@tedsuo which change? This adds parentOrElse which is already defined in specs, that's what I commented to remove the default change

**Comment by tedsuo** [comment]
Sorry, that was the bit I was referring to, that's for splitting it out.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/config/TraceConfig.java (hunk 2 above)
Comment (MitchellDumovic) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/config/TraceConfig.java]: I added an issue to the spec repo to discuss this: https://github.com/open-telemetry/opentelemetry-specification/issues/728 cc @anuraaga 

One thing that I admit to being confused about is how the sampler is meant to interact with the `TraceFlags`, which is slightly inconsistent across languages. If anyone can clear up some of the inconsistencies for me that would be helpful.
